src/platform/webusb/device.rs

```diff
@@ -0,0 +1,307 @@
+use std::{sync::Arc, time::Duration};
+
+use wasm_bindgen_futures::{js_sys::Array, wasm_bindgen::JsCast, JsFuture};
+use web_sys::{
+    js_sys::Uint8Array, UsbControlTransferParameters, UsbDevice, UsbInTransferResult,
+    UsbOutTransferResult,
+};
+
+use crate::{
+    descriptors::{validate_config_descriptor, DESCRIPTOR_TYPE_CONFIGURATION},
+    transfer::{web_to_nusb_status, Control, EndpointType, TransferError, TransferHandle},
+    DeviceInfo, Error,
+};
+
+#[derive(Clone)]
+pub(crate) struct WebusbDevice {
+    pub device: UsbDevice,
+    config_descriptors: Vec<Vec<u8>>,
+}
+
+/// SAFETY: This is NOT safe at all.
+unsafe impl Sync for WebusbDevice {}
+unsafe impl Send for WebusbDevice {}
+
+impl WebusbDevice {
+    pub(crate) async fn from_device_info(d: &DeviceInfo) -> Result<Arc<WebusbDevice>, Error> {
+        let window = web_sys::window().unwrap();
+        let navigator = window.navigator();
+        let usb = navigator.usb();
+        let devices = JsFuture::from(usb.get_devices()).await.unwrap();
+        let devices: Array = JsCast::unchecked_from_js(devices);
+
+        for device in devices {
+            let device: UsbDevice = JsCast::unchecked_from_js(device);
+            if device.eq(&d.device) {
+                JsFuture::from(device.open()).await.unwrap();
+
+                let config_descriptors = extract_decriptors(&device).await?;
+
+                #[allow(clippy::arc_with_non_send_sync)]
+                return Ok(Arc::new(Self {
+                    device,
+                    config_descriptors,
+                }));
+            }
+        }
+        Err(Error::other("device not found"))
+    }
+
+    // pub(crate) fn handle_events(&self) {
+    //     todo!()
+    // }
+
+    pub(crate) fn configuration_descriptors(&self) -> impl Iterator<Item = &[u8]> {
+        self.config_descriptors.iter().map(|d| &d[..])
+    }
+
+    pub(crate) fn active_configuration_value(&self) -> u8 {
+        self.device
+            .configuration()
+            .map(|c| c.configuration_value())
+            .unwrap_or_default()
```

Comment:
@kevinmehall is this an okay assumption? Judging from the current function signature I would assume that the fallback is 0. In Rust fashion we could change the interface or keep it as is if it is non-critical.

Reply:
Yes, configuration 0 means "not configured" in the underlying GET_CONFIGURATION response as well.

Reply:
Thanks! Sorry my USB knowledge is very limited and finding stuff in the spec is sometimes a bit tricky =)